[inductor] Make {freezing,ir}.py pass follow-imports typechecking

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113722
* #113721
* #113720
* #113610
* __->__ #113534

I used a couple of type-ignore comments in ir.py because it constructs
short-lived instances of FixedLayout and GraphModuleSerializer, just to
call a single method on them that doesn't use all their members. Making
those unused members optional would make the rest of the code a lot
messier with sprinkled `assert` statements.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

int3 committed:
[inductor] Make {freezing,ir}.py pass follow-imports typechecking

I used a couple of type-ignore comments in ir.py because it constructs
short-lived instances of FixedLayout and GraphModuleSerializer, just to
call a single method on them that doesn't use all their members. Making
those unused members optional would make the rest of the code a lot
messier with sprinkled `assert` statements.

[ghstack-poisoned]

int3 commented:
The one test failure looks spurious, doesn't replicate on stacked diffs

desertfire commented on torch/_tensor.py:
```diff
@@ -81,6 +81,7 @@ def _rebuild_from_type_v2(func, new_type, args, state):
 class Tensor(torch._C.TensorBase):
     _dynamo_static_input_type: str  # Set in `mark_static_address`
     _dynamo_dynamic_indices: Set[int]  # Set in `mark_dynamic`
+    _is_param: bool  # Set on torch.nn.Parameter instances
```
What is the reason for adding this?

int3 replied:
It's set here: 

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/freezing.py#L154

and here:

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/freezing.py#L169

voznesenskym commented on torch/_inductor/freezing.py:
```diff
@@ -77,9 +77,9 @@ def freeze(
-    if torch._guards.TracingContext.get():
-        fw_metadata = torch._guards.TracingContext.get().fw_metadata
-        params_flat = torch._guards.TracingContext.get().params_flat
+    if tracing_context := torch._guards.TracingContext.get():
```
walrus operator smelly -

int3 replied:
hm, is this a codebase convention or a personal preference? :p

voznesenskym commented on torch/_tensor.py:
```diff
@@ -81,6 +81,7 @@ def _rebuild_from_type_v2(func, new_type, args, state):
 class Tensor(torch._C.TensorBase):
     _dynamo_static_input_type: str  # Set in `mark_static_address`
     _dynamo_dynamic_indices: Set[int]  # Set in `mark_dynamic`
+    _is_param: bool  # Set on torch.nn.Parameter instances
```
This is a change to a core type in pytorch eager, cc @albanD

int3 replied:
Note that this syntax doesn't have any runtime effect. The attribute is already being set / read in a number of places; this just makes it explicit for mypy.

albanD replied:
But the point is that this attribute shouldn't exist for regular Tensors. It is not even used for the plain Tensor class.
So adding this here will make the `__annotations__` wrong as this field should NOT be read on a `Tensor` object.

int3 replied:
It looks like this code is reading it on Tensor objects:

https://github.com/pytorch/pytorch/blob/79e3833703c421cb9cb9ec2c45751a09abbe6e9b/torch/nn/parameter.py#L6-L10

int3 committed:
Update on "[inductor] Make {freezing,ir}.py pass follow-imports typechecking"


I used a couple of type-ignore comments in ir.py because it constructs
short-lived instances of FixedLayout and GraphModuleSerializer, just to
call a single method on them that doesn't use all their members. Making
those unused members optional would make the rest of the code a lot
messier with sprinkled `assert` statements.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Make {freezing,ir}.py pass follow-imports typechecking"


I used a couple of type-ignore comments in ir.py because it constructs
short-lived instances of FixedLayout and GraphModuleSerializer, just to
call a single method on them that doesn't use all their members. Making
those unused members optional would make the rest of the code a lot
messier with sprinkled `assert` statements.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
Removed attribute on Tensor, used type-ignore comments instead

int3 committed:
Update on "[inductor] Make {freezing,ir}.py pass follow-imports typechecking"


I used a couple of type-ignore comments in ir.py because it constructs
short-lived instances of FixedLayout and GraphModuleSerializer, just to
call a single method on them that doesn't use all their members. Making
those unused members optional would make the rest of the code a lot
messier with sprinkled `assert` statements.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

albanD reviewed (approved):
Sounds good!
